packages/alfa-table/src/cell.ts

```diff
@@ -285,19 +285,22 @@ export namespace Cell {
       explicitHeaders?: Iterable<Element>;
       implicitHeaders?: Iterable<Element>;
     }): Builder {
-      return Builder.of(
+      this._cell = Cell.of(
         kind,
         x,
         y,
         width,
         height,
         element,
         variant,
-        downwardGrowing,
-        scope,
-        explicitHeaders,
-        implicitHeaders
+        List.empty()
       );
+      this._downwardGrowing = downwardGrowing;
+      this._scope = scope;
+      this._explicitHeaders = List.from(explicitHeaders);
+      this._implicitHeaders = List.from(implicitHeaders);
+
+      return this; // for chaining
```

Comment:
Why break immutability here?

Reply:
Cells (the concept) de facto mutate during the build process (notably adding the scope on header cells).
When `Table#slots` was returning an `Option<Cell>` that was OK since "mutating" the cell could just be done by mutating `Table#slots` instead.
But now, we need `Table#slots` to return a `Set<Cells>` to accept these table model errors. When the cell change, we would need to remove the old one from the set, and add the new one, which is more complex than just updating the cell.

OTOH, I agree than only the `Cell.Builder` should mutate, the product should be immutable 😓

Reply:
If preserving immutability can be done simply be removing the old cell and adding the new one then let's do that. Shortcutting this at the cost of losing immutability seems like the wrong tradeoff.

Reply:
I'm afraid that "simply" is not really going to be simple…
There's many places where cells are created and updated, often in bulk by sub-steps, and it's not that easy to track what the "old" cell is. The easiest way I see is to add an `id` to each cell that is kept through all incarnations and use this for comparing.

Reply:
This really sounds like something we should fix regardless then as it seems more complex than it has to be 🤔 First things first then: Why are the cells stored in a set? With a set, we need to worry about keeping a reference to the old cell. If we were using a list, we'd simply identify cells by their index, which is effectively the ID you're describing.

Reply:
I don't see an easy way to fetch the index of the old version, unless I already have the old version and then whatever the collection, I can just `reject` the old one.
I've initially chose `Set` precisely to let it handle the "no duplicate" part for me.

Reply:
The point was that you'd always use the same index for the "same" cell and then simply update the cell at a given index. Either way, there seems to be some fundamental architectural hurdles in this package that we need to sort out before we pile on additional complexity.

Reply:
Currently, cell mutation happens deeply at various places, so it's not easy to track where the cell comes from and what was its position in the list for when we need to update this list.
Additionally, a given cell might be present in several of these sets (if it covers several slots, which AFAIU must be the case for slots that are covered multiple time, i.e. the case where a set is needed rather than an option), potentially at different positions each time. So passing the information of where it was in each seems pretty much over complicated also.

Reply:
For posterity, this is what I had in mind:

https://github.com/Siteimprove/alfa/blob/4c5828ab8e9a74cc50f86760adc2982d91424106/packages/alfa-table/src/table.ts#L105-L106

This way, cells are stored in an array and referenced in the table grid by their index. When updating cells, they're simply overwritten in the cell array:

https://github.com/Siteimprove/alfa/blob/4c5828ab8e9a74cc50f86760adc2982d91424106/packages/alfa-table/src/table.ts#L527-L533

Header assignment is done at the very last step by mapping over the cell array and assigning headers:

https://github.com/Siteimprove/alfa/blob/4c5828ab8e9a74cc50f86760adc2982d91424106/packages/alfa-table/src/table.ts#L190-L192